File: packages/shared-types/api_schemas.ts
/**
 * packages/shared-types/api_schemas.ts
 * Shared TypeScript request/response types for every WorkLine AI endpoint.
 * Import from here in api.ts and all frontend stores.
 *
 * Keep in sync with api_schemas.py (Python side).
 */

// ── Block / Node / Edge ───────────────────────────────────────────────────

export interface NodePosition {
    x: number;
    y: number;
}

export interface NodeData {
    label: string;
    config?: Record<string, unknown>;
    reasoning?: string;
}

export interface WorkflowNode {
    id: string;
    type: string;
    position: NodePosition;
    data: NodeData;
    reasoning?: string;
}

export interface WorkflowEdge {
    id: string;
    source: string;
    target: string;
    edge_type?: 'default' | 'condition_true' | 'condition_false';
}

// ── Planning ──────────────────────────────────────────────────────────────

export interface PlanRequest {
    goal: string;
    conversation_id?: number;
}

export interface WorkflowProposal {
    title: string;
    reasoning: string;
    nodes: WorkflowNode[];
    edges: WorkflowEdge[];
    conversation_id: number;
}

export interface ConversationTurn {
    id: number;
    role: 'user' | 'assistant';
    content: string;
    proposal_json?: WorkflowProposal;
    created_at: string;
}

export interface Conversation {
    id: number;
    org_id?: number;
    workflow_id?: number;
    created_at: string;
    turns: ConversationTurn[];
}

// ── Workflows ─────────────────────────────────────────────────────────────

export interface WorkflowCreateRequest {
    name: string;
    description?: string;
    nodes: WorkflowNode[];
    edges: WorkflowEdge[];
}

export interface Workflow {
    id: number;
    org_id?: number;
    name: string;
    description?: string;
    status: 'draft' | 'active' | 'archived';
    version: number;
    parent_version_id?: number;
    created_at: string;
}

export interface WorkflowDetail extends Workflow {
    nodes: WorkflowNode[];
    edges: WorkflowEdge[];
}

// ── Domain Packs ──────────────────────────────────────────────────────────

export interface DomainPack {
    name: string;
    status: 'installed' | 'available';
}

// ── Auth ──────────────────────────────────────────────────────────────────

export interface LoginRequest {
    email: string;
    password: string;
}

export interface TokenResponse {
    access_token: string;
    token_type: string;
}

// ── Runs ──────────────────────────────────────────────────────────────────

export interface WorkflowRun {
    id: number;
    workflow_id: number;
    status: 'pending' | 'running' | 'completed' | 'failed' | 'awaiting_review';
    started_at: string;
    ended_at?: string;
    logs?: unknown;
}

export interface RunTriggerResponse {
    status: string;
    mode: 'background' | 'synchronous';
    task_id?: string;
    result?: unknown;
}

File: packages/shared_types/api_schemas.py
"""
packages/shared-types/api_schemas.py
Shared Pydantic request/response models for every WorkLine AI endpoint.
Both the API routers and any test clients should import from here.
"""
from datetime import datetime
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, ConfigDict


# ── Block schemas ──────────────────────────────────────────────────────────

class NodePosition(BaseModel):
    x: float
    y: float


class NodeData(BaseModel):
    label: str
    config: Dict[str, Any] = {}
    reasoning: Optional[str] = None


class NodeSchema(BaseModel):
    id: str
    type: str
    position: NodePosition
    data: NodeData
    reasoning: Optional[str] = None


class EdgeSchema(BaseModel):
    id: str
    source: str
    target: str
    edge_type: str = "default"   # default | condition_true | condition_false


# ── Planning ───────────────────────────────────────────────────────────────

class PlanRequest(BaseModel):
    goal: str
    conversation_id: Optional[int] = None


class WorkflowProposal(BaseModel):
    title: str
    reasoning: str
    nodes: List[NodeSchema]
    edges: List[EdgeSchema]
    conversation_id: int


class ConversationTurnOut(BaseModel):
    id: int
    role: str                            # "user" | "assistant"
    content: str
    proposal_json: Optional[Dict[str, Any]] = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ConversationOut(BaseModel):
    id: int
    org_id: Optional[int] = None
    workflow_id: Optional[int] = None
    created_at: datetime
    turns: List[ConversationTurnOut] = []

    model_config = ConfigDict(from_attributes=True)


# ── Workflows ──────────────────────────────────────────────────────────────

class WorkflowBase(BaseModel):
    name: str
    description: Optional[str] = None


class WorkflowCreate(WorkflowBase):
    nodes: List[NodeSchema] = []
    edges: List[EdgeSchema] = []


class WorkflowResponse(WorkflowBase):
    id: int
    org_id: Optional[int] = None
    status: str
    version: int
    parent_version_id: Optional[int] = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class WorkflowDetailResponse(WorkflowResponse):
    nodes: List[Dict[str, Any]] = []
    edges: List[Dict[str, Any]] = []


# ── Auth ───────────────────────────────────────────────────────────────────

class LoginRequest(BaseModel):
    email: str
    password: str


class TokenResponse(BaseModel):
    access_token: str
    token_type: str = "bearer"


# ── Runs ───────────────────────────────────────────────────────────────────

class WorkflowRunResponse(BaseModel):
    id: int
    workflow_id: int
    status: str
    started_at: datetime
    ended_at: Optional[datetime] = None
    logs: Optional[Any] = None

    model_config = ConfigDict(from_attributes=True)
# ── Domain Packs ──────────────────────────────────────────────────────────

class DomainPackResponse(BaseModel):
    name: str # mechanical, etc
    status: str # installed | available


# ── Dashboard ─────────────────────────────────────────────────────────────

class DashboardSummary(BaseModel):
    total_runs_week: int
    success_rate: float
    avg_duration: float
    active_drift_alerts: int


class RecentRun(BaseModel):
    id: int
    workflow_name: str
    triggered_by: str
    status: str
    started_at: datetime
    duration: float


class DashboardRecentRuns(BaseModel):
    runs: List[RecentRun]


class DashboardDriftAlert(BaseModel):
    id: int
    workflow_name: str
    metric: str
    baseline_val: float
    current_val: float
    created_at: datetime


class DashboardDriftAlerts(BaseModel):
    alerts: List[DashboardDriftAlert]
